dev/set_matrix.py

```diff
@@ -0,0 +1,412 @@
+    ['foo']
+    >>> process_requirements({"== 0.1": ["foo"], "== 0.2": ["bar"]}, "0.2")
+    ['bar']
+    >>> process_requirements({"== dev": ["foo"]}, "0.1")
+    []
+    >>> process_requirements({"< dev": ["foo"]}, "0.1")
+    ['foo']
+    >>> process_requirements({"> 0.1": ["foo"]}, "dev")
+    ['foo']
+    >>> process_requirements({"== dev": ["foo"]}, "dev")
+    ['foo']
+    >>> process_requirements({"> 0.1, != dev": ["foo"]}, "dev")
+    []
+    """
+    if requirements is None:
+        return []
+
+    if isinstance(requirements, list):
+        return requirements
+
+    dev_numeric = "9999.9999.9999"
+
+    if version == "dev":
+        version = dev_numeric
+
+    if isinstance(requirements, dict):
+        for ver_spec, packages in requirements.items():
+            op_and_ver_pairs = map(get_operator_and_version, ver_spec.split(","))
+            match_all = all(
+                op(LooseVersion(version), LooseVersion(dev_numeric if ver == "dev" else ver))
+                for op, ver in op_and_ver_pairs
+            )
+            if match_all:
+                return packages
+        else:
+            return []
+
+    raise TypeError("Should not reach here")
+
+
+def remove_comments(s):
+    return "\n".join(l for l in s.strip().split("\n") if not l.strip().startswith("#"))
+
+
+def main():
+    args = parse_args()
+
+    print(divider("Parameters"))
+    print(json.dumps(vars(args), indent=2))
+
+    print(divider("Log"))
+    config = read_yaml(CONFIG_PATH)
+    try:
+        config_ref = read_yaml(args.ref_config)
+    except Exception as e:
+        print("Failed to read '{}' due to: '{}'".format(args.ref_config, e))
+        config_ref = {}
+
+    flavors = set(x.split("-")[0] for x in config.keys())
+    changed_flavors = get_changed_flavors(args.changed_files, flavors)
+
+    job_names = []
+    includes = []
+    should_include_all_items = not args.diff_only
+
+    for flavor, cfgs in config.items():
+        package_info = cfgs.pop("package_info")
+
+        should_include_all_items_in_this_flavor = (
+            should_include_all_items
+            or any(flavor.startswith(x) for x in changed_flavors)
+            or (flavor not in config_ref)
+            or (package_info != config_ref[flavor]["package_info"])
+        )
+
+        for key, cfg in cfgs.items():
+            if (
+                should_include_all_items_in_this_flavor
+                or (key not in config_ref[flavor])
+                or (cfg != config_ref[flavor][key])
+            ):
+                print("Processing {}.{}".format(flavor, key))
+
+                # released versions
+                all_versions = get_released_versions(package_info["pip_release"])
+                versions = filter_versions(
+                    all_versions, cfg["minimum"], cfg["maximum"], cfg.get("unsupported"),
+                )
+                versions = select_latest_micro_versions(versions)
```

Comment:
Should we **always** include teh latest version in each flavor in the matrxi to replace the current flavor tests (`run-python-flavor-tests.sh` and `run-python-tf-tests.sh`)?

Reply:
Is there a limit on how many CI jobs can run concurrently? Will the aggregate test time be decreased? Assuming there isn't an issue with having a slightly larger CI job count, I like this idea. It simplifies the steps taken by each job and should make debugging easier.

Could we do this as a follow-up PR?

Reply:
Yes, there is a limit. It's 20: https://docs.github.com/en/free-pro-team@latest/actions/reference/usage-limits-billing-and-administration#usage-limits

Reply:
Got it. One other potential complicating factor is that `run-python-flavor-tests` includes test coverage for things like AzureML & our top-level `mlflow.autolog()` API (which requires that many ML libraries be installed at once). Perhaps we can revisit this some time in the future.